File: backends/fastapi/app/core/security.py
"""鉴权原语：bcrypt 密码哈希 + JWT access token + 不透明 refresh token。"""

from __future__ import annotations

import base64
import hashlib
import secrets
import time
import uuid

import bcrypt
import jwt

from app.core.config import settings


# ---------- 密码 ----------
def hash_password(plain: str) -> str:
    salt = bcrypt.gensalt(rounds=settings.bcrypt_rounds)
    return bcrypt.hashpw(plain.encode("utf-8"), salt).decode("utf-8")


def verify_password(plain: str, hashed: str) -> bool:
    try:
        return bcrypt.checkpw(plain.encode("utf-8"), hashed.encode("utf-8"))
    except ValueError:
        return False


# ---------- Access Token (JWT HS256) ----------
def create_access_token(*, user_id: uuid.UUID, nickname: str, avatar_id: str) -> tuple[str, int]:
    """返回 (token, 过期秒数)。"""
    now = int(time.time())
    exp = now + settings.access_token_ttl_seconds
    payload = {
        "sub": str(user_id),
        "nickname": nickname,
        "avatarId": avatar_id,
        "iat": now,
        "exp": exp,
    }
    token = jwt.encode(payload, settings.jwt_secret, algorithm=settings.jwt_algorithm)
    return token, settings.access_token_ttl_seconds


class AccessTokenError(Exception):
    def __init__(self, reason: str):
        super().__init__(reason)
        self.reason = reason


def decode_access_token(token: str) -> dict:
    try:
        return jwt.decode(token, settings.jwt_secret, algorithms=[settings.jwt_algorithm])
    except jwt.ExpiredSignatureError as e:
        raise AccessTokenError("access_token_expired") from e
    except jwt.InvalidTokenError as e:
        raise AccessTokenError("invalid_token") from e


# ---------- Refresh Token ----------
# 不透明：256-bit 随机 + base64url；DB 存 sha256 hash，比对时 hash 输入。
# 请求侧的无效 token 也参与 hash 后自然查无匹配，不能因为编码问题变成 500。
def generate_refresh_token() -> str:
    raw = secrets.token_bytes(32)
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def hash_refresh_token(raw: str) -> str:
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()

File: fullstacks/next/src/lib/server/security.ts
/**
 * 鉴权原语：bcrypt 密码哈希 + JWT access token + 不透明 refresh token。
 *
 * 与 FastAPI 参考实现的行为对齐：
 * - access token：JWT HS256，1h，payload 含 sub/nickname/avatarId/iat/exp
 * - refresh token：256-bit 随机 → base64url（不透明），DB 存 sha256
 */
import "server-only";

import bcrypt from "bcryptjs";
import { randomBytes, createHash } from "node:crypto";
import { SignJWT, jwtVerify } from "jose";
import { env } from "../env";

const secret = new TextEncoder().encode(env.jwtSecret);

export async function hashPassword(plain: string): Promise<string> {
  const salt = await bcrypt.genSalt(env.bcryptRounds);
  return bcrypt.hash(plain, salt);
}

export async function verifyPassword(plain: string, hashed: string): Promise<boolean> {
  try {
    return await bcrypt.compare(plain, hashed);
  } catch {
    return false;
  }
}

export interface AccessTokenClaims {
  sub: string;
  nickname: string;
  avatarId: string;
  iat: number;
  exp: number;
}

export async function createAccessToken(payload: {
  userId: string;
  nickname: string;
  avatarId: string;
}): Promise<{ token: string; expiresIn: number }> {
  const now = Math.floor(Date.now() / 1000);
  const exp = now + env.accessTokenTtlSeconds;
  const token = await new SignJWT({
    nickname: payload.nickname,
    avatarId: payload.avatarId,
  })
    .setProtectedHeader({ alg: "HS256" })
    .setSubject(payload.userId)
    .setIssuedAt(now)
    .setExpirationTime(exp)
    .sign(secret);
  return { token, expiresIn: env.accessTokenTtlSeconds };
}

export class AccessTokenError extends Error {
  reason: "access_token_expired" | "invalid_token";
  constructor(reason: "access_token_expired" | "invalid_token") {
    super(reason);
    this.reason = reason;
  }
}

export async function decodeAccessToken(token: string): Promise<AccessTokenClaims> {
  try {
    const { payload } = await jwtVerify(token, secret, { algorithms: ["HS256"] });
    return payload as unknown as AccessTokenClaims;
  } catch (e) {
    const msg = (e as Error).message ?? "";
    if (/expired/i.test(msg)) throw new AccessTokenError("access_token_expired");
    throw new AccessTokenError("invalid_token");
  }
}

// ---------- Refresh Token ----------
export function generateRefreshToken(): string {
  return randomBytes(32).toString("base64url");
}

export function hashRefreshToken(raw: string): string {
  return createHash("sha256").update(raw, "utf8").digest("hex");
}
